cv32e40s/tests/programs/custom/zc_test/zc_test.S

```diff
@@ -0,0 +1,1165 @@
+**                                        https://solderpad.org/licenses/SHL-2.1/
+** Unless required by applicable law or agreed to in writing, any work distributed under the License is distributed on
+** an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the License for the
+** specific language governing permissions and limitations under the License.
+*******************************************************************************
+**
+** Assembly functions to help the tests in the zc_tests.c source file.
+**
+*******************************************************************************
+*/
+
+.section .data
+glb_irq_line:
+  .word 0
+glb_irq_delay:
+  .word 0
+stored_ra:
+  .word 0
+
+.section .text
+
+.global glb_irq_line
+.global glb_irq_delay
+
+// Functions
+.global enable_all_irq
+
+//Generated Functions
+.global interrupt_push_pop
+.global interrupt_popret
+.global interrupt_popretz
+.global interrupt_mvsa
+.global interrupt_mvas
+
+
+//instruction tags
+.global push_0
+.global pop_0
+.global push_1
+.global pop_1
+.global push_2
+.global pop_2
+.global push_3
+.global pop_3
+.global push_4
+.global pop_4
+.global push_5
+.global pop_5
+.global push_6
+.global pop_6
+.global push_7
+.global pop_7
+.global push_8
+.global pop_8
+.global push_9
+.global pop_9
+.global push_10
+.global pop_10
+.global push_11
+.global pop_11
+.global popret_0
+.global popret_1
+.global popret_2
+.global popret_3
+.global popret_4
+.global popret_5
+.global popret_6
+.global popret_7
+.global popret_8
+.global popret_9
+.global popret_10
+.global popret_11
+.global popretz_0
+.global popretz_1
+.global popretz_2
+.global popretz_3
+.global popretz_4
+.global popretz_5
+.global popretz_6
+.global popretz_7
+.global popretz_8
+.global popretz_9
+.global popretz_10
+.global popretz_11
+.global mvsa_0
+.global mvsa_1
+.global mvsa_2
+.global mvsa_3
+.global mvsa_4
+.global mvsa_5
+.global mvsa_6
+.global mvsa_7
+.global mvsa_8
+.global mvsa_9
+.global mvsa_10
+.global mvsa_11
+.global mvsa_12
+.global mvsa_13
+.global mvsa_14
+.global mvsa_15
+.global mvas_0
+.global mvas_1
+.global mvas_2
+.global mvas_3
+.global mvas_4
+.global mvas_5
+.global mvas_6
+.global mvas_7
+.global mvas_8
+.global mvas_9
+.global mvas_10
+.global mvas_11
+.global mvas_12
+.global mvas_13
+.global mvas_14
+.global mvas_15
+
+// Funtions
+enable_all_irq:
+  //set all fields of mie, hardwired are ignored (WARL)
+  li	t0, 0xFFFFFFFF
+	 csrrw x0, mie, t0
+
+  li	t0, 0x00000008
+	 csrrs x0, mstatus, t0
+
+  jalr	x0, 0(ra)
+
+trigger_irq:
+  // trigger irq
+  li  t0, 0x00800140
+  la  t1, glb_irq_line
+  lw  t2, 0(t1)
+  la  t1, glb_irq_delay
+  lw  t3, 0(t1)
+  sw  t2, 0(t0)
+  sw  t3, 4(t0)
+
+  //return to caller
+  jalr	x0, 0(ra)
+
+
+
+//Generated Functions
+interrupt_push_pop:
+  la t0, stored_ra
+  sw ra, 0(t0)
+
+
+  pp_case_4:
+    li t1, 4
+    bne a0, t1, pp_case_5
+    jal ra, trigger_irq
+  push_0:
+    cm.push {ra}, -64
+    nop
```

Comment:
what are the purpose of these nops?

Reply:
To let the interrupts be taken in an easy to estimate(and debug) program flow.